user/caribou/module/src/DSO9254AEventConverter.cc

```diff
@@ -0,0 +1,377 @@
+    //                             might end up having the wrong one. 
+    // Finn: I am not sure if I actually solved this issue! Lets discuss.    
+    rawdata.resize( sizeof(datablock[0]) * datablock.size() / sizeof(uintptr_t) );
+    std::memcpy(&rawdata[0], &datablock[0], sizeof(datablock[0]) * datablock.size() );
+
+    
+    // needed per event
+    uint64_t block_words;
+    uint64_t pream_words;
+    uint64_t chann_words;
+    uint64_t block_posit = 0;
+    uint64_t sgmnt_count = 1;
+
+    
+    // loop 4 channels
+    for( int nch = 0; nch < 4; nch++ ){      
+
+      std::cout << "Reading channel " << nch << std::endl;
+
+      // Obtaining Data Stucture
+      block_words = rawdata[block_posit + 0];
+      pream_words = rawdata[block_posit + 1];
+      chann_words = rawdata[block_posit + 2 + pream_words];
+      std::cout << "  " << block_posit << " current position in block\n";
+      std::cout << "  " << block_words << " words per segment\n";
+      std::cout << "  " << pream_words << " words per preamble\n";
+      std::cout << "  " << chann_words << " words per channel data\n";
+      std::cout << "  " << rawdata.size() << " size of data rawdata\n";
+           
+
+      // read preamble:
+      std::string preamble = "";
+      for(int i=block_posit + 2;
+              i<block_posit + 2 + pream_words;
+              i++ ){
+        preamble += (char)rawdata[i];
+      }
+      LOG(DEBUG) << preamble;
+      std::cout << "Preamble: \n  " << preamble << std::endl;  
+      // parse preamble to vector of strings
+      std::string s;
+      std::vector<std::string> vals;
+      std::istringstream stream(preamble);
+      while(std::getline(stream,s,',')){
+        vals.push_back(s);
+      }
+      
+
+      // getting timestamps from preamble which is rather imprecise
+      timestamp = timeConverter( vals[15], vals[16] );
+
+      
+      // Pick needed preamble elements
+      np.push_back( stoi( vals[2]) );
+      dx.push_back( stod( vals[4]) * 1e9 );
+      x0.push_back( stod( vals[5]) * 1e9 );
+      dy.push_back( stod( vals[7]) );
+      y0.push_back( stod( vals[8]) );
+      if( vals.size() == 25 ) {// this is segmented mode, possibly more than one waveform in block
+	sgmnt_count = stoi( vals[24] );
+      }
+      int points_per_words = np.at(nch)/(chann_words/sgmnt_count);
+      if( np.at(nch)%chann_words/sgmnt_count ) LOG(WARNING) << "incomplete waveform";
+      /*
+      std::cout << "  dx " << dx.at(nch) << std::endl
+		<< "  x0 " << x0.at(nch) << std::endl
+		<< "  dy " << dy.at(nch) << std::endl
+		<< "  y0 " << y0.at(nch) << std::endl;
+      */
+
+      // need once per channel
+      std::vector<double> ped;
+      std::vector<double> amp;
+      std::vector<std::vector<double>> waves;
+      
+      
+      for( int s=0; s<sgmnt_count; s++){ // loop semgents
+
+     	// container for one waveform
+	std::vector<double> wave;
+	
+	// prepare histogram with corresponding binning
```

Comment:
Shouldn't these only be prepared if `generateRoot` is true? Now we create them even if we don't want to write out to ROOT file...

Reply:
yes